src/handlers/permissions-revoke.ts

```diff
@@ -0,0 +1,115 @@
+import type { BaseMessageReply } from '../core/message-reply.js';
+import type { DidResolver } from '../did/did-resolver.js';
+import type { EventLog } from '../types/event-log.js';
+import type { MessageStore } from '../types/message-store.js';
+import type { MethodHandler } from '../types/method-handler.js';
+import type { PermissionsGrantMessage, PermissionsRevokeMessage } from '../types/permissions-types.js';
+
+import { authenticate } from '../core/auth.js';
+import { messageReplyFromError } from '../core/message-reply.js';
+import { PermissionsRevoke } from '../interfaces/permissions-revoke.js';
+import { DwnInterfaceName, DwnMethodName, Message } from '../core/message.js';
+
+export class PermissionsRevokeHandler implements MethodHandler {
+  constructor(private didResolver: DidResolver, private messageStore: MessageStore, private eventLog: EventLog) { }
+
+  public async handle({
+    tenant,
+    message
+  }: { tenant: string, message: PermissionsRevokeMessage }): Promise<BaseMessageReply> {
+    let permissionsRevoke: PermissionsRevoke;
+    try {
+      permissionsRevoke = await PermissionsRevoke.parse(message);
+    } catch (e) {
+      return messageReplyFromError(e, 400);
+    }
+
+
+    // Authentication
+    try {
+      await authenticate(message.authorization, this.didResolver);
+    } catch (e) {
+      return messageReplyFromError(e, 401);
+    }
+
+    // Authorization
+    const permissionsGrantId = message.descriptor.permissionsGrantId;
+    const permissionsGrantMessage = await this.messageStore.get(
+      tenant,
+      permissionsGrantId
+    ) as PermissionsGrantMessage | undefined;
+
+    if (permissionsGrantMessage === undefined) {
+      return {
+        status: {
+          code   : 400,
+          detail : `Could not find PermissionsGrant with CID ${permissionsGrantId}`
+        }
+      };
+    }
+
+    try {
+      await permissionsRevoke.authorize(permissionsGrantMessage);
+    } catch (e) {
+      return messageReplyFromError(e, 401);
+    }
+
+    // Revoke must have `dateCreated` after that of the grant
+    if (message.descriptor.dateCreated < permissionsGrantMessage.descriptor.dateCreated) {
+      return {
+        status: { code: 400, detail: 'PermissionsRevoke has earlier date than associated PermissionsGrant' }
+      };
+    }
+
+    // query for previous revocations of this grant
+    const query = {
+      interface : DwnInterfaceName.Permissions,
+      method    : DwnMethodName.Revoke,
+      permissionsGrantId,
+    };
+    const existingRevokesForGrant: PermissionsRevokeMessage[] = await this.messageStore.query(tenant, query) as PermissionsRevokeMessage[];
+
+    // Conflict 409 if the grant already has an older revoke
```

Comment:
I imagine this will be the one of the most surprising/nuanced pieces of functionality.

Consider the scenario where Alice sends a `PermissionsRevoke` to her local DWN, then a moment later sends a `PermissionsRevoke` to her remote DWN for the same `PermissionsGrant`. Upon DWN sync, we should use the earlier `PermissionsRevoke` and delete/bounce the later `PermissionsRevoke`.

Reply:
Thanks! Super helpful!